### PR TITLE
Implement reduce_accumulators_per_key in SparkRDDOperations (# 41)

### LINKED ISSUE #41: Using accumulators in SparkOperations
# Context

Currently the development of [Accumulator Framework](https://github.com/OpenMined/PipelineDP/issues/37) is in progress. The idea is that **Accumulator Framework** is independent of the pipeline framework. This issue is about implementing usage of **Accumulator Framework** with Spark.

The interface of accumulators:

```
class Accumulator(ABC):
     def merge(cls,  accumulators: Iterable[Accumulator]) -> Accumulator
     def add(v)
     def compute_metrics()
```

# Goals
To implement the following method of [SparkRDDOperations](https://github.com/OpenMined/PipelineDP/blob/65d00ee6c906a36b8a173ddbb83d7e3bf71905b4/pipeline_dp/pipeline_operations.py#L102):

`def reduce_accumulators_per_key(self, col):`

where **col** contains elements:  (key, accumulator) 

and the return is a collection of elements (key, accumulator), such that all elements per each key are merged.

# Notes
It looks like **reduce_accumulators_per_key** might be implement with usage of [reduce_by_key](https://spark.apache.org/docs/latest/api/python/reference/api/pyspark.RDD.reduceByKey.html).

### PR DESCRIPTION
**Pull request by boyaryn**
## Description
Implementation of accumulator reduction for Spark processing engine.
Resolves #41.

## Affected Dependencies
N/A

## How has this been tested?
- Tests are to be added as they need the [Accumulator Framework](https://github.com/OpenMined/PipelineDP/issues/37) which is to be defined.

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [x] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [ ] My changes are covered by tests

### CODE REVIEW

File: pipeline_dp/pipeline_operations.py [hunk 1]
```diff
@@ -157,6 +161,19 @@ def count_per_element(self, rdd, stage_name: str = None):
         return rdd.map(lambda x: (x, 1))\
             .reduceByKey(lambda x, y: (x + y))
 
+    def reduce_accumulators_per_key(self, rdd, stage_name: str = None):
+        """Reduce the input RDD so that all elements per each key are merged.
```
Comment (dvadym) [inline pipeline_dp/pipeline_operations.py:165]: Please move this comment to a base class (just replacing, RDD -> collections). The idea that the code that uses PipelineOperations don't need to know which framework they use

File: pipeline_dp/pipeline_operations.py [hunk 2]
```diff
@@ -157,6 +161,19 @@ def count_per_element(self, rdd, stage_name: str = None):
...
+
+        Args:
+          rdd: input rdd which contains tuples (key, accumulator)
+          stage_name: name of the stage
+
+        Returns:
+          An RDD of tuples (key, accumulator).
+
+        """
+        rdd.reduceByKey(lambda acc1, acc2: acc1.merge([acc2]))
```
Comment (dvadym) [inline pipeline_dp/pipeline_operations.py:175]: let's assume for now that merge takes a list (then let's discuss on accumulator framework how to allow acceptance of 1 accumulator)

**Review by dvadym — CHANGES_REQUESTED** [review]
Thanks, it looks good! 

Could you please add unit tests? You can use some simple mock object. Like Accumulator that does counting

File: pipeline_dp/pipeline_operations.py (hunk 1 above)
Comment (boyaryn) [reply pipeline_dp/pipeline_operations.py]: Good idea. Updated.

File: pipeline_dp/pipeline_operations.py (hunk 2 above)
Comment (boyaryn) [reply pipeline_dp/pipeline_operations.py:175]: okay

**Comment by boyaryn** [comment]
> Thanks, it looks good!
> 
> Could you please add unit tests? You can use some simple mock object. Like Accumulator that does counting

I'll add unit tests. Is it better to wait when the Accumulator interface is available in `main` so that the tests can check real-world cases?

**Comment by chinmayshah99** [comment]
> I'll add unit tests. Is it better to wait when the Accumulator interface is available in `main` so that the tests can check real-world cases?
Let's add some mock tests for now as @dvadym suggested. Once the `Accumulator` interface is available, we can move to more real-world test cases.

**Comment by dvadym** [comment]
I agree, please create very a simple MockAccumulator (eg. CounterAccumulator). Anyway in unittests it's better to use mock objects that real accumulators (which will be pretty complicated objects, since they will have support of Differential privacy)

**Comment by dvadym** [comment]
FYI: [SumAccumulator] Mock accumulator (https://github.com/OpenMined/PipelineDP/blob/e33a7c724179dba77b7646597f2af80e47ffe5a7/tests/pipeline_operations_test.py#L316) was implemented for testing of BeamOperations. 

Please use SumAccumulator for testing

File: pipeline_dp/pipeline_operations.py [hunk 3]
```diff
@@ -254,8 +253,8 @@ def count_per_element(self, rdd, stage_name: str = None):
         return rdd.map(lambda x: (x, 1))\
             .reduceByKey(lambda x, y: (x + y))
 
-    def reduce_accumulators_per_key(self, col, stage_name: str = None):
-        raise NotImplementedError()
+    def reduce_accumulators_per_key(self, rdd, stage_name: str = None):
+        return rdd.reduceByKey(lambda acc1, acc2: acc1.merge([acc2]))
```
Comment (dvadym) [inline pipeline_dp/pipeline_operations.py:257]: acc1.merge -> acc1.add_accumulator(acc2)

**Review by dvadym — CHANGES_REQUESTED** [review]
Thanks, it looks good. Just a minor fix.

**Review by dvadym — APPROVED** [review]
Thanks for contributing!

**Pull request merged**